kernel/sched.c

```diff
@@ -121,6 +125,53 @@ int32_t z_sched_prio_cmp(struct k_thread *thread_1,
 		return (int32_t) (d2 - d1);
 	}
 #endif
+
+#ifdef CONFIG_TIMESLICING
+	/*
+	 * We want to avoid this scenario:
+	 *
+	 *        CPU0         CPU1
+	 * t0 +----------+ .          .
+	 * t1 |          | |          |
+	 * t2 | thread A | +----------+
+	 * t3 |          | |          |
+	 * t4 +----------+ | thread B | <-- Timeslice done: thread A requeued.
+	 * t5 |          | |          | <-- Some event: thread D queued.
+	 * t6 | thread C | +----------+ <-- Timeslice done: thread B requeued
+	 * t7 |          | |          |     and thread A in front of the queue.
+	 * t8 +----------+ | thread A |
+	 * t9 |          | |          |
+	 * .. .          . +----------+
+	 *
+	 * To be fair, the scheduler should have preferred D over A at t6
+	 * but A quickly re-entered the queue when its timeslice expired.
+	 * in order to help fairness, we apply a priority bias against threads
+	 * with a newly expired timeslice for a duration equivalent to another
+	 * timeslice. This is less likely but this applies on !SMP as well.
```

Comment:
This bit I'm mostly just not understanding.  This seems wrong to me.  The scheduler selects the highest priority thread that has been waiting the longest; it's a simple rule and that's how we document it (in more or less exactly those words).

When A's timeslice expired, it went in the back of the queue.  D became runnable only later, when A had already been waiting.  Why exactly shouldn't A run first?  If the app wants D to run in preference to threads being timesliced, it should give it a higher priority.  Otherwise it needs to get in line and wait for its turn like everyone else.

This seems like an attempt at an I/O-based "interactivity" heuristic scheduling algorithm, where threads that become runnable get higher priorities than ones that have been running for a while (the presumption being that they woke up due to user interaction from some device, and that the long-running stuff is probably "in the background").  That's indeed how it works in many big kernels.  But that's not what we have in Zephyr, which is essentially "POSIX realtime plus cooperative priorities".

Let's remove this patch and test case.  If we want to bring this back, let's do it as a separate feature under its own kconfig (CONFIG_SCHED_INTERACTIVE) or something, maybe as (net) another threshold level where the bottom tier of preemptivle priorities have this "depress priority if they've been running" behavior.